readme: add initial storage updates (#188)
hashicorp/nomad-driver-virt: README.md

Nomad Virt Driver
==================
The virt driver task plugin expands the types of workloads Nomad can run to add virtual machines. Currently

> leveraging the power of libvirt, the virt drivers allows users to define virtual machine tasks using the 

Reviewer: Either "driver allows" or "drivers allow" -- since you changed drivers to plural, I suppose we want "allow"?

```suggestion
leveraging the power of libvirt, the virt drivers allow users to define virtual machine tasks using the 
```

Buuut it is only one driver, even if it has multiple providers...

Nomad job spec.

> **_IMPORTANT:_** This plugin is in tech preview, still under active development, there might be breaking changes in future releases

**: This is an Alpha version still under development**

## Features

* Use the job's `task.config` to define the virtual machine (VM).
* Start/stop virtual machines.
* [Nomad runtime environment](https://www.nomadproject.io/docs/runtime/environment.html) is populated.
* Use Nomad alloc data in the virtual machine.
* Publish ports.
* Monitor the memory consumption.
* Monitor CPU usage.
* Task config cpu value is used to populate virtual machine CpuShares.
* The tasks `task`, `alloc`, and `secrets` directories are mounted within the VM at the filesystem
  root. These are currently mounted read-only to prevent excessive amounts of data being written to 

>   the host filesystem. Please see the [filesystem concepts page](https://developer.hashicorp.com/nomad/docs/concepts/filesystem) 

Reviewer: This seems a bit strange to me, but I'll assume that this is currently the case, and the doc matches reality.

Just one trailing space dangling:

```suggestion
  root. These are currently mounted read-only to prevent excessive amounts of data being written to 
  the host filesystem. Please see the [filesystem concepts page](https://developer.hashicorp.com/nomad/docs/concepts/filesystem)
```

Author: What seems strange, the read-only mounts?

Reviewer: Yeah, it's not like most OSes cram a bunch of stuff into these particular dirs. If OSes had any special expectations for them, it would be risky for us to mount on them.

Author: Sure, it's less an issue of the guest system and more an issue of the task that is running within the guest system. A task is limited to filling up its volume(s), but if these host directories were mounted read/write, there's nothing to prevent the task from writing to them and filling the host system.

Reviewer 2: Note that's the case for all other task drivers today, which is a shame but hard to fix in a backwards compatible way. I'd _really_ like to do this but it'd almost certainly require moving `template` blocks into a sidecar task.

https://github.com/hashicorp/nomad-driver-virt/issues/69 has a discussion of the reasoning for this particular driver.

Reviewer: To clarify, when you say "that's the case" for other task drivers, you mean that these dirs _are_ writeable.
* https://developer.hashicorp.com/nomad/docs/reference/runtime-environment-settings#task-directories

all say ~
> can be used to store arbitrary/secret data

There are other ways to share data between tasks, like volumes with `multi-node-*` capabilities, but `secrets/` is pretty special, being a tmpfs only in memory, and anyway writing to these dirs is a long-standing pattern.

I'm not sure why issue 69 was closed, as @jrasell said there

> will keep this issue open as we certainly want to remove this restriction in the future.

But this is all tangential to the goal of this PR :yum:

      }

      config {
        default_user_password = "password"
        cmds                  = ["python3 -m http.server 8000"]

        disk {
          size   = "10GiB"
          source {

>             image  = "local/focal-server-cloudimg-amd64.img"

Reviewer: More stray spaces. There are some down in network_interface, too, but we haven't gotten to networking yet.

```suggestion
        disk {
          size = "10GiB"
          source {
            image = "local/focal-server-cloudimg-amd64.img"
```

## Driver Configuration

* **image_paths** - Host paths containing image files allowed to be used by tasks.
* **provider** - Named block containing provider configuration. Defaults to libvirt.
* **storage_pools** - Block containing storage pool configuration.

### Provider - libvirt

* **password** - The libvirt password to use for authentication.
* **uri** - The libvirt driver to use. Defaults to `qemu:///system`.
* **user** - The libvirt user to use for authentication.

### Storage pools 

Storage pools contain volumes which are created for, and attached to, task VMs. Two 
types of storage pools are supported by the driver: directory and Ceph. Directory
storage pools are host local storage pools with volumes stored at a specified path.

> Ceph storage pools are RBD based volumes stored in Ceph. 

Reviewer: Quite a few more trailing spaces around here.

I'll stop commenting on them individually, and instead just point at:

```
grep -n ' $' README.md
```

Author: Thanks for pointing this out. It seems like trailing spaces aren't getting auto-clipped when I'm in markdown mode.

It is required that a default storage pool is assigned. If the configuration only

> defines a single storage pool, that storage pool is automatically the default. 

Reviewer: I won't ding all the instances of passive voice, but this one felt particularly strained.

```suggestion
A default storage pool must be assigned. If the configuration only
defines a single storage pool, that storage pool is automatically the default. 
```

Still passive (assigned by whom?), but more direct, IMO.

* **ceph** - Named block containing Ceph based storage pool configuration.
* **default** - Name of the default storage pool. If only one storage pool is defined, it is automatically the default.
* **directory** - Named block containing directory based storage pool configuration.

#### Storage pool - directory 

* **path** - Host path to contain the pool volumes.

#### Storage pool - ceph 

* **authentication** - Block containing authentication configuration.
  * **username** - Ceph client name .
  * **secret** - Ceph client key (base64 encoded).
* **hosts** - List of Ceph monitors.
* **pool** - Name of the Ceph pool.

### Examples

Minimal configuration defining a directory storage pool on the host and defining
a directory for image files which tasks may reference:

``` hcl
plugin "nomad-driver-virt" {
  config {
    image_paths = ["/var/lib/virt/images"]
    storage_pools {
      directory "local-storage" {
        path = "/var/lib/virt/storage"
      }
    }
  }
}
```

Full configuration example with libvirt configuration, multiple host directories
for image files, and two storage pools (one directory storage pool and one Ceph 

> storage pool) with the directory storage pool marked as the default storage pool: 

Reviewer: ```suggestion
This full libvirt configuration example has a username and password and allows
multiple host directories for image files. It defines two storage pools, one
directory and one Ceph, and the directory storage pool is marked as the default: 
```

I wish it were easy in github to reflow the line breaks 🤷

``` hcl
plugin "nomad-driver-virt" {
  config {
    provider "libvirt" {
      uri      = "qemu:///system"
      user     = "libvirt-username"
      password = "libvirt-pass"
    }
    
    image_paths = [
      "/var/lib/virt/images",
      "/opt/custom-images",
    ]
    
    storage_pools {
      default = "local-storage"
      directory "local-storage" {
        path = "/var/lib/virt/storage"
      }
      
      ceph "remote-storage" {

>         pool  = "nomad-pool"
>         hosts = [

Reviewer: Super nit-picky spacing. In general, multiline assignments (lists, maps, blocks) don't want alignment from prior lines.

```suggestion
        pool = "nomad-pool"
        hosts = [
```

* **bus_type** - Bus type for the disk. Defaults to `virtio`.
* **chained** - Disk is an overlay on the source.
* **devname** - Device name used within the VM. Auto-generated by default.
* **driver** - Driver to use for the disk. Usage and default value is provider specific.
* **format** - Format of the disk. Default is provider specific.
* **kind** - Kind of disk defined. Defaults to `disk`.
* **pool** - Storage pool to place volume created from this definition. Defaults to the default storage pool.
* **primary** - Disk is the primary to boot the VM.
* **read_only** - Disk is read only.
* **size** - Size of the disk as bytes, or string (example: `20GB` or `15GiB`)
* **sparse** - Disk should be sparsely populated.
* **source** - Block containing disk source configuration.
  * **format** - Format of the image. Auto-detected if unset.
  * **image** - Image to write to the disk. Overwrites any existing information on disk.
  * **volume** - Volume in storage pool to clone.
* **volume** - Nomad volume to back the disk.

#### Example 

The example below shows the task and disk configuration to define a primary disk in the directory storage pool and a secondary

> empty disk in the Ceph storage pool:

Reviewer: Remind me: the first disk is implicitly primary? I don't see how it's being specified here otherwise.

Also, how is the second disk in the Ceph storage pool? They're both `pool = "local-storage"`...

Author: If only a single disk is defined, it will automatically be assigned as primary.

As for the second disk being in the Ceph pool: through the magic of typos.

```hcl
job "python-server" {
  group "virt-group" {
    task "virt-task" {
      driver = "virt"

>       artifact {

Reviewer: nit: I always put `driver = ""` right above the `config{}` block, as they're tightly coupled.

        source      = "http://cloud-images.ubuntu.com/focal/current/focal-server-cloudimg-amd64.img"
        destination = "local/focal-server-cloudimg-amd64.img"
        mode        = "file"
      }

      config {
        disk {
          size   = "10GiB"
          pool   = "local-storage"

>           source {

Reviewer: ```suggestion
          size = "10GiB"
          pool = "local-storage"
          source {
```

            image = "local/focal-server-cloudimg-amd64.img"
          }
        }

        disk {
          size = "20GB"
          pool = "local-storage"
        }
      }
    }
  }
}

```

If the storage pool already contains a volume with the focal server image, it can be cloned to remove the need of
downloading and applying the image. Once the volume is cloned, it will be automatically resized to the requested
size:

```hcl
job "python-server" {
  group "virt-group" {
    task "virt-task" {
      driver = "virt"

      config {
        disk {

>           size   = "10GiB"
>           pool   = "local-storage"
>           source {

Reviewer: ```suggestion
          size = "10GiB"
          pool = "local-storage"
          source {
```

            volume = "focal-server-cloudimg-amd64.img"
          }
        }
      }
    }
  }
}

```

Instead of making a full clone of the source volume, a chained copy may be created which overlays the new volume
on the source volume creating a copy-on-write volume:

```hcl
job "python-server" {
  group "virt-group" {
    task "virt-task" {
      driver = "virt"

      config {
        disk {
          size    = "10GiB"
          pool    = "local-storage"
          chained = true
          source {
            volume = "focal-server-cloudimg-amd64.img"
          }
        }
      }
    }
  }
}

```

Chained copies may also be used when providing a source image. A new volume will be created for the image and

> any tasks that define a chained disk with that source image will be chained to that volume:

Reviewer: "with that source image" -- This is tracked by a hash of the image, right? I wonder if saying so here might be helpful for users.

Author: Currently yes but that may not (and likely will not) always be the case. Might be worthwhile to add that information after that stabilizes a bit.

```hcl
job "python-server" {
  group "virt-group" {
    task "virt-task" {
      driver = "virt"

      artifact {
        source      = "http://cloud-images.ubuntu.com/focal/current/focal-server-cloudimg-amd64.img"

>         destination = "local/focal-server-cloudimg-amd64.img"

Reviewer: This example seems a little odd, as the source image will be downloaded every time, even though no new volume is needed on a second pass?

Am I interpreting that right? I'm not sure what I would suggest as an efficient pattern here...

Author: The efficient pattern here will be the addition of handling downloads directly when the value of the `source.image` is a URL.

        mode        = "file"
      }

      config {
        disk {
          size    = "10GiB"
          pool    = "local-storage"
          chained = true
          source {
            image = "local/focal-server-cloudimg-amd64.img"
          }
        }
      }
    }
  }
}

```

### Network Configuration

The following configuration options are available within the task's driver configuration block:

* **bridge** - Block configuration for connecting to a bridged network.

>   * **name** - Name of the bridge interface to use. Defaults to `virbr0` which is the default libvirt network.

Reviewer: I think `name` is required, with no default?
